arrows/vxl/aligned_edge_detection.cxx

```diff
@@ -0,0 +1,304 @@
+      {
+        if( val_j >= grad_j( i, j - 1 ) && val_j >= grad_j( i, j + 1 ) )
+        {
+          output( i, j, 1 ) = static_cast< OutputType >( val_j );
+        }
+      }
+    }
+  }
+  return output;
+}
+
+// ----------------------------------------------------------------------------
+// Calculate potential edges
+template < typename PixType, typename GradientType >
+vil_image_view< PixType >
+aligned_edge_detection::priv
+::calculate_aligned_edges( vil_image_view< PixType > const& input,
+                           vil_image_view< GradientType >& grad_i,
+                           vil_image_view< GradientType >& grad_j )
+{
+  // Calculate sobel approx in x/y directions
+  vil_sobel_3x3( input, grad_i, grad_j );
+
+  // Take absolute value of gradients
+  vil_transform< GradientType, GradientType( GradientType ) >(
+    grad_i, std::abs );
+  vil_transform< GradientType, GradientType( GradientType ) >(
+    grad_j, std::abs );
+
+  // Perform NMS in vertical/horizonal directions and threshold magnitude
+  vil_image_view< PixType > output =
+    nonmax_suppression< GradientType, PixType >( grad_i, grad_j );
+  return output;
+}
+
+// ----------------------------------------------------------------------------
+template < typename pix_t >
+vil_image_view< pix_t >
+aligned_edge_detection::priv
+::filter( vil_image_view< pix_t > const& input_image )
+{
+  auto const source_ni = input_image.ni();
+  auto const source_nj = input_image.nj();
+
+  vil_image_view< float > grad_i{ source_ni, source_nj };
+  vil_image_view< float > grad_j{ source_ni, source_nj };
+
+  vil_image_view< pix_t > aligned_edges =
+    calculate_aligned_edges< pix_t, float >( input_image, grad_i, grad_j );
+
+  // Perform extra op if enabled
+  if( produce_joint_output )
+  {
+    vil_image_view< pix_t > combined_edges{ source_ni, source_nj };
+    combined_edges.fill( 0 );
+
+    // Add vertical and horizontal edge planes together and smooth
+    vil_image_view< pix_t > joint_nms_edges;
+
+    vil_math_image_sum( vil_plane( aligned_edges, 0 ),
+                        vil_plane( aligned_edges, 1 ),
+                        joint_nms_edges );
+
+    unsigned half_step = smoothing_half_step;
+    unsigned min_dim = std::min( joint_nms_edges.ni(), joint_nms_edges.nj() );
+
+    if( 2 * half_step + 1 >= min_dim )
+    {
+      half_step = ( min_dim - 1 ) / 2;
+    }
+
+    if( half_step != 0 )
+    {
+      vil_gauss_filter_2d( joint_nms_edges, combined_edges, smoothing_sigma,
+                           half_step );
+    }
+    else
+    {
+      vil_copy_reformat( joint_nms_edges, combined_edges );
+    }
+    vil_image_view< pix_t > all_channels( joint_nms_edges.ni(),
+                                          joint_nms_edges.nj(), 3 );
+    vil_plane( all_channels, 0 ).deep_copy( vil_plane( aligned_edges, 0 ) );
+    vil_plane( all_channels, 1 ).deep_copy( vil_plane( aligned_edges, 1 ) );
+    vil_plane( all_channels, 2 ).deep_copy( joint_nms_edges );
```

Comment:
The source of the prior bug was I was using assignment (shallow copy) rather than deep copy. However, I'm not thrilled about this current approach since it seems to waste memory. I was curious if there's any better way to deal with this?

Reply:
More generally, is there a cheap channel-wise concatenation operation in `vil` that can be used in similar situations?

Reply:
I don't think so? If I am understanding this correctly, you are combining planes from separate images into one image. There is not a way to do that that is generally usable (maybe in very specific instances sticking entirely to VXL, but not generally) that doesn't involve a deep copy, because the resulting image needs to have all the data in a single memory block, and the input data is in separate memory blocks. Plus, even if you could, the resulting image would need to refcount the inputs.

What you *might* be able to do, but would at least require some refactoring, would be to allocate the result image up front and do the other operations in place on "slices" of that image.

BTW, what happens to `combined_edges`? Is it me, or does it just get discarded?

Reply:
The commit seems to have been messed up by the force push. The following code sets it by reference.

`
if( half_step != 0 )
{
    vil_gauss_filter_2d( joint_nms_edges, combined_edges, smoothing_sigma,
                         half_step );
}
else
{
    vil_copy_reformat( joint_nms_edges, combined_edges );
}
`

Reply:
This is probably not important to do unless it proves to be a bottleneck, but you can avoid the deep copy with refactoring as @mwoehlke-kitware suggested.  This is why many of the `vil` functions are written to return the output image in a parameter passed in by reference.  This allows the caller to decide which memory the results are placed in.

Reply:
Despite Matt W. pointing it out, I still made a mistake and the images were never smoothed. In updating this to be more efficient, I fixed that issue.